presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java

```diff
@@ -0,0 +1,398 @@
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.union;
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.defaultPrefixes;
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.isTablesEnumeratingTable;
+import static io.prestosql.metadata.MetadataListing.listSchemas;
+import static io.prestosql.metadata.MetadataListing.listTableColumns;
+import static io.prestosql.metadata.MetadataListing.listTablePrivileges;
+import static io.prestosql.metadata.MetadataListing.listTables;
+import static io.prestosql.metadata.MetadataListing.listViews;
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static io.prestosql.spi.type.TypeUtils.writeNativeValue;
+import static java.util.Objects.requireNonNull;
+
+public class InformationSchemaPageSource
+        implements ConnectorPageSource
+{
+    private final Session session;
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+
+    private final String catalogName;
+    private final InformationSchemaTable table;
+    private final Iterator<QualifiedTablePrefix> prefixIterator;
+    private final OptionalLong limit;
+
+    private final List<Type> types;
+    private final Queue<Page> pages = new ArrayDeque<>();
+    private final PageBuilder pageBuilder;
+    private long recordCount;
+
+    private final Function<Page, Page> projection;
+
+    private long completedBytes;
+    private long memoryUsageBytes;
+    private boolean closed;
+
+    public InformationSchemaPageSource(
+            Session session,
+            Metadata metadata,
+            AccessControl accessControl,
+            InformationSchemaTableHandle tableHandle,
+            List<ColumnHandle> columns)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        requireNonNull(tableHandle, "tableHandle is null");
+        requireNonNull(columns, "columns is null");
+
+        catalogName = tableHandle.getCatalogName();
+        table = tableHandle.getTable();
+        Set<QualifiedTablePrefix> prefixes = tableHandle.getPrefixes();
+        if (isTablesEnumeratingTable(table)) {
+            if (prefixes.equals(defaultPrefixes(catalogName))) {
+                prefixes = metadata.listSchemaNames(session, catalogName).stream()
```

Comment:
Please postpone this until the moment when it is actually needed. Maybe you could use some `Supplier<Iterator<QualifiedTablePrefix>>` for prefix iterator supplier or custom `QualifiedTablePrefixIterator`